idseq_dag/util/s3.py

```diff
@@ -203,6 +228,11 @@ def fetch_from_s3(src,  # pylint: disable=dangerous-default-value
                             named_args=named_args
                         )
                     )
+                if abspath.startswith(config["REF_DIR"]):
+                    os.makedirs(config["REF_FETCH_LOG_DIR"], exist_ok=True)
+                    with open(os.path.join(config["REF_FETCH_LOG_DIR"], abspath_hash), "w") as fh:
```

Comment:
We may want to protect this section with a multiprocessing lock as well

Reply:
No, this line is also protected by the lock on line 139/151 (https://github.com/chanzuckerberg/idseq-dag/pull/218/files#diff-30956c749673657f32d6e09a0ac2e49dR151).

I am also puzzled by this design pattern of putting locks everywhere. Generally speaking, that's what leads to deadlocks. It is preferable to structure operations in such a way that they either share nothing, or move data into place atomically or over such a short period of time that a race is implausible. That way, even if they end up racing against each other, the processes will produce the same result.

Reply:
The current design heavily relies on multiprocessing, semaphore and locks, which makes it hard to follow the logic. I hadn't noticed those locks you mentioned and I agree they should be enough to protect this code section.

Reply:
Hmm, I'm reviewing it again and I believe those locks may not be enough. They are related to a specific destination file. Therefore, if two different processes are downloading different files, this code section will be unprotected from concurrency issues.

Reply:
What are the concurrency issues that you're worried about?

Reply:
I just realized you are creating a json file for each destination file. You are right, that lock should prevent any concurrency issues.

Reply:
We mostly rely on multithreading for concurrency.   We hardly ever use multiprocessing, and in such a limited way that it would be rather straightforward to replace with actual subcommands.  We literally never communicate back from child to parent, except through a file.  We use multiprocessing locks because they are safe for both multithreading and multiprocessing.  Unfortunately, the same cannot be said about some library modules, particularly the python logging package, which has often been a source of issues with multiprocessing.  For this reason, I would be quite open to removing the use of multiprocess start/end and replacing with subcommands.  Shouldn't take more than a day or two.

Reply:
In Python, I prefer to use concurrent.futures.ThreadPoolExecutor for concurrency, and nothing else. That guides developers into a particular programming model that is far safer than dealing with locks individually. While I have no doubt that you can write threadsafe code using basic locking primitives, I don't trust myself to do so, and I don't think developers in general will be able to safely change your code.

With that said, I don't think we should be improving this codebase any further; instead we should focus on replacing it with the other alternatives we discussed (like miniwdl) that share nothing and use concurrency on a lower level. This discussion only strengthens and underscores my conviction that we should avoid investing in this codebase and retire it as quickly as possible, loading and running code in the individual steps in https://github.com/chanzuckerberg/idseq-dag/tree/master/idseq_dag/steps from an external executor as a migration strategy.

Reply:
I am open to using the pattern you recommend, but unclear on how it relates to the pattern when multiple of these threads wish to download the same file, and something in your infrastructure ought to ensure the file is downloaded just once.   Clearly that requires some kind of locking, if threads are used.

Separately, and more importantly, even if our plan is to completely deprecate this infrastructure and switch to something better, I think you would agree with me that in the meantime we should at least not make it less stable.   I see you agree because you decided to create private boto sessions for every thread.  That's wonderful.